common/lib/common-definitions/src/logger.ts

```diff
@@ -31,6 +40,13 @@ export interface ITelemetryBaseEvent extends ITelemetryProperties {
  * Implemented by hosting app / loader
  */
 export interface ITelemetryBaseLogger {
+    /**
+     * An optional boolean which indicates to the logger that tags (i.e. `ITaggedTelemetryPropertyType` objects) are
+     * in use. Eventually this will be a required property, but this is a stopgap that allows older hosts to
+     * continue to pass through telemetry without trouble (this property will simply show up undefined), while
+     * our current logger impl in `telmetry-utils` handles tags in a separate manner.
+     */
+    supportsTags?: true;
```

Comment:
It feels like full cycle is too long:
1. Add optional supportsTags
2. Make it required
3. Eventually remove it (as it does not have any value after some time).

Consider instead adding new method (sendTagged?) that takes tagged or not input, and make send() take non-tagged input. Make both optional and let code that uses this interface (hopefully only TelemetryLogger class) to deal with both (depending on which one is present).

Over time, make sendTagged required and remove send.

You can even skip step and require sendTagged right away, but internally ensure that we can cope with it missing (by ensuring that all of our code compiles correctly when it's marked optional). That way you skip step and require host to move to new API, while retaining back compat for as long as we need, and making last (removal of send()) step without breaking compat. It also removes ambiguity, everything.

This is not required, simply sharing how I'd approach it :)

Reply:
Dunno, feels identical but then you're left with a function sendTagged whose name I don't like.

I do wish there were a nice way to make supportsTags required to implement (so they say false at first) but optional on our side, but can't think of a good way.

Reply:
Ok I thought of a plan I like better:

- On the updated interface, make `supportsTags: true`. When hosts pick up this release they have to support tags right away, but it's not hard because they can mostly copy/paste our code
- In telemetry-utils, define a type `IOldTelemetryBaseLogger` which is a copy of the interface before this change. Have `ITelemetryLogger` extend a union of the new and old types. Now we are properly modeling how this will play out in terms of what capabilities a base logger will have.
- Everywhere else is fine to just use the new logger type because the tagging code is the only place the distinction actually matters.
- Distant future: remove the old type and handling for tags if `supportsTags` is missing.

If modifying `ITelemetryLogger`'s base class doesn't work (I'm skeptical having thought it through a bit more) then maybe we just do some casting in the tagging code. But what I like about this is it's way simpler and more explicit for consumers.

Reply:
Ok against better judgment I've been mulling this over in my head this morning. I had to delete my last comment because it made no sense, ha. But I do see how Vlad's idea is less error-prone. I'm going to stop thinking about this and let you all wrap it up next week. Bye!

Reply:
I think I'll keep `supportsTags?` instead of going the `sendTagged` route. In my mind, it's better to keep the only method in `ITelemetryBaseLogger` simply named `send()` (and then any future subtlety can be hashed out with types or comments). Hypothetically, if we used Vlad's plan but still wanted to change the name of the new method back to `send` at some point after we've made tags required, then that would get fairly messy right?